examples/MQTT/mqtt_basic/mqtt_basic.ino

```diff
@@ -64,6 +64,7 @@ void connect() {
     Serial.print(".");
     if (millis() - clTimeout > 5001) {
       Serial.println();
+      mesh.renewAddress();
       return;
     }
```

Comment:
#### My main concern is ignoring the return value for `mesh.renewAddress()`
Should this be a `do {...} while()` instead, like we used in the mesh examples' `setup()`?
```cpp
  Serial.print("connecting...");
  uint32_t clTimeout = millis() + 5001; // pre-calculate timeout for performance
  while (!client.connect(clientID)) {
    do {
      Serial.println(".");
    while (mesh.renewAddress() == MESH_DEFAULT_ADDRESS && millis() < clTimeout)
    if (millis() > clTimeout) {
      return;
    }
```
I imagine `Ethernet.update()` (called after this) needs an active mesh connection. I'm not sure how `MQTTClient::connect()` behaves. Is the lost connection implicit via MQTT client's "keep alive" behavior?

Or maybe it would be better to have this `::connect()` return a bool?
```cpp
    if (millis() - clTimeout > 5001) {
      Serial.println();
      return mesh.renewAddress() != MESH_DEFAULT_ADDRESS;
    }

    // ... then at end of function body
    return true;
    // or if compile size isn't a major concern here:
    return mesh.mesh_address != MESH_DEFAULT_ADDRESS
```
So, `loop()` can act more intuitively.


I also noticed that below this is
```cpp
    if (!mesh.checkConnection()) {
      if (mesh.renewAddress() == MESH_DEFAULT_ADDRESS) {
        mesh.begin();
      }
    }
```
Although, I'm assuming that was meant for your hot-swapping radios tactic.

---
PS - I'm also noticing timeouts calculated in loop iterations. I know this can't be avoided in `loop()`, but here (in `::connect()`), the timeouts are local vars.

Reply:
a: I'd rather just stick with this behaviour. Its simple and tries to renew the address, if that fails it takes another 5 seconds of connection timeout before trying again. This helps alleviate traffic on the network when there is a bunch of interference and/or the mesh needs to reconverge.
b: I think so. I think we could put a lot more thought into these examples, but for now I was just intending to enhance its behaviour lol. I should probably include a non-blocking connect example as well at some point. Give me some time.
c & d: See b

Reply:
Also to note, the mesh.begin() call I now use to switch channels on the radio network too. I have a failover node setup with MQTT and the network automatically fails over to another channel if interference is detected ongoing for a certain period of time and picks up a new channel. I can also switch it over manually with the click of a button. Has been invaluable for testing and troubleshooting. Pretty neat too!

Reply:
[`PubSubClient::connect()`](https://github.com/knolleary/pubsubclient/blob/2d228f2f862a95846c65a8518c79f48dfc8f188c/src/PubSubClient.cpp#L181) does ping the broker. According to MQTT protocol, the connection state is determined by ping's response.

In the lib's [listed limitations](https://github.com/knolleary/pubsubclient/tree/master#limitations):
> The keepalive interval is set to 15 seconds by default. This is configurable via `MQTT_KEEPALIVE` in `PubSubClient.h` or can be changed by calling `PubSubClient::setKeepAlive(keepAlive)`.

Reply:
I found problems with using the pubsub library years ago, so switched to the https://github.com/256dpi/arduino-mqtt library.


> Set more advanced options:
> 
> void setKeepAlive(int keepAlive);
> void setCleanSession(bool cleanSession);
> void setTimeout(int timeout);
> void setOptions(int keepAlive, bool cleanSession, int timeout);
> The keepAlive option controls the keep alive interval in seconds (default: 10).
> The cleanSession option controls the session retention on the broker side (default: true).
> The timeout option controls the default timeout for all commands in milliseconds (default: 1000).

Reply:
Its a little hard to tell, but that lib's `connected()` seems to be more implicit (depending on a proper init of network connection). Its an interesting lib that seems much better suited to embedded hardware.

Reply:
Yeah, the pubsub library is pretty commonly used it seems, but there is a note 
> The library cannot currently be used with hardware based on the ENC28J60 chip

So I'm not sure if its compatibility issues with the chip or the hardware but I found instability when using pubsub that wasn't there with the MQTT lib. It might be fixed at this point, but I haven't experimented with pubsub for a long time now.

BTW, just committed some changes, still need to test the new example if I get some time, but 99% sure it will work.